### PR TITLE
[tcgc] SdkTypes Design Doc

### PR DESCRIPTION
**Pull request by iscai-msft**

### CODE REVIEW

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 1]
```diff
@@ -0,0 +1,1477 @@
...
+getWidgetOperationStatus is Operations.GetResourceOperationStatus<Widget>;
+
+// Widget Operations
+/** Creates or updates a Widget asynchronously */
+@pollingOperation(getWidgetOperationStatus)
+createOrUpdateWidget is Operations.LongRunningResourceCreateOrUpdate<Widget>;
+```
+
+```ts
+const sdkPagingServiceMethod = {
```
Comment (tadelesh) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:630]:
```suggestion
const serializeLroServiceMethod = {
```

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 2]
```diff
@@ -0,0 +1,1477 @@
...
+    ...
+  }
+}
+```
+
+### SdkServiceMethod
+
+An `SdkServiceMethod` is any service method on a client that calls the service.
+
+The actual service call is a separate property on the method, `.operation`. This way, our service methods are able to abstract away the protocol used to call the service (i.e. `http` or `gRPC`)
```
Comment (tadelesh) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:400]: Currently for Python adoption, we do not relay on `SdkServiceMethod` info. Do we have any plan to leverage it?

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 3]
```diff
@@ -0,0 +1,1477 @@
...
+
+They all share this base:
+
+```ts
+interface SdkTypeBase {
+  // the raw TypeSpec type. Some of our types are currently
+  // created by tcgc. Those won't have an original type
+  __raw?: Type;
+  kind: string;
+  nullable: boolean;
```
Comment (tadelesh) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:1006]: If a type is reference type, we could not set `nullable` to the type, instead need to set it to the property/parameter. We need to revisit the design of `nullable`.
```
model Test {
}
model A {
  prop: Test | null;
}
model B {
  prop: Test;
}

```

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 4]
```diff
@@ -0,0 +1,1477 @@
...
+@tag("Widgets")
+interface Widgets {
+  @get list(): Widget[] | Error;
+  @get read(@path id: string): Widget | Error;
+  @post create(...Widget): Widget | Error;
+  @patch update(...Widget): Widget | Error;
+  @delete delete(@path id: string): void | Error;
+  @route("{id}/analyze") @post analyze(@path id: string): string | Error;
+}
+```
```
Comment (tadelesh) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:55]: Maybe provide the whole yaml file link of `sdkPackage` for this example will be much intuitionistic?

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 5]
```diff
@@ -0,0 +1,1477 @@
...
+> {
+  __raw?: Operation;
+  name: string;
+  access: AccessFlags | undefined;
+  parameters: SdkParameter[];
+  apiVersions: string[];
+  description?: string;
+  details?: string;
+  overloads?: SdkMethod<TServiceOperation>[];
+  overloading?: SdkMethod<TServiceOperation>;
```
Comment (tadelesh) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:359]: what are these for?

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 6]
```diff
@@ -0,0 +1,1478 @@
...
+  | SdkPathParameter
+  | SdkHeaderParameter
+  | SdkBodyParameter;
+```
+
+```ts
+interface SdkModelPropertyTypeBase {
+  __raw?: ModelProperty;
+  type: SdkType;
+  nameInClient: string;
```
Comment (iscai-msft) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:808]: change to `name`

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 7]
```diff
@@ -0,0 +1,1478 @@
...
+```
+
+#### SdkMethodParameter
+
+Represents a parameter to a client method. Does not have any transport (i.e. HTTP) related information on it. This is solely meant to represent information that we expect sdk users to pass in. We then take care to map the method input and create our request to the service using the information users have inputted.
+
+```ts
+export interface SdkMethodParameter extends SdkModelPropertyTypeBase {
+  kind: "method";
+}
```
Comment (iscai-msft) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:858]: include section for query / header params etc

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 8]
```diff
@@ -0,0 +1,1478 @@
...
+interface SdkArrayType extends SdkTypeBase {
+  kind: "array";
+  valueType: SdkType;
+}
+```
+
+### SdkTupleType
+
+```ts
+interface SdkTupleType extends SdkTypeBase {
```
Comment (iscai-msft) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:1069]: see if we can get rid of tuple type

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 9]
```diff
@@ -0,0 +1,1478 @@
...
+```
+
+### SdkEnumType
+
+```ts
+export interface SdkEnumType extends SdkTypeBase {
+  kind: "enum";
+  name: string;
+  // If it's an anonymous enum, this is the name we generate
+  generatedName?: string;
```
Comment (iscai-msft) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:1092]:
```ts
interface SdkEnumType {
  name: string; // always exists, can be generated;
  generatedName: boolean; // true if name is generated, false if not
}
```

Comment (iscai-msft) [reply docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:1092]: should have `details?` as well

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 10]
```diff
@@ -0,0 +1,1478 @@
...
+
+### SdkModelType
+
+```ts
+export interface SdkModelType extends SdkTypeBase {
+  kind: "model";
+  // purposely can also be header / query params for fidelity with TypeSpec
+  properties: SdkModelPropertyType[];
+  name: string;
+  isFormDataType: boolean;
```
Comment (iscai-msft) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:1148]: should we put this in usage?

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx [hunk 11]
```diff
@@ -0,0 +1,1478 @@
...
+  isError: boolean;
+  generatedName?: string;
+  description?: string;
+  details?: string;
+  access?: AccessFlags;
+  usage: UsageFlags;
+  additionalProperties?: SdkType;
+  discriminatorValue?: string;
```
Comment (iscai-msft) [inline docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:1156]: should we make a separate `discriminatedModel` type with the discriminator goop on it?

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx (hunk 2 above)
Comment (iscai-msft) [reply docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:538]: we should leverage `SdkServiceMethod` for creating the method signature, eventually we should fully move over to this structure bc it gives us much more flexibility. Our hands are kind of tied right now with trying to maintain the old Python SdkType structure

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx (hunk 3 above)
Comment (iscai-msft) [reply docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:1144]: Yeah, I agree with you. This was a discussion at some point in time, but I think nullable should actually be on the property / parameter instead. I'm going to update the design to this

File: docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx (hunk 4 above)
Comment (iscai-msft) [reply docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx:55]: I'm not sure what you mean by this comment, I have updated to add a hierarchy of types, hopefully this fixes this comment as well

**Review by jhendrixMSFT — APPROVED** [review]

**Pull request merged**